core/src/main/java/hudson/ClassicPluginStrategy.java

```diff
@@ -566,23 +566,30 @@ private static void unzipExceptClasses(File archive, File destDir, Project prj)
     /**
      * Used to load classes from dependency plugins.
      */
-    final class DependencyClassLoader extends ClassLoader {
+    static final class DependencyClassLoader extends ClassLoader {
         /**
          * This classloader is created for this plugin. Useful during debugging.
          */
         private final File _for;
 
         private List<Dependency> dependencies;
 
+        private final PluginManager pluginManager;
+
         /**
-         * Topologically sorted list of transient dependencies.
+         * Topologically sorted list of transient dependencies. Lazily initialized via double-checked locking.
```

Comment:
Indeed, I think you would always wind up calculating it. I do not know why Kohsuke opted for lazy initialization here, nor do I know why he called it "transient dependencies" (I think it should read "transitive"). Anyway, there is a lot of code that could be improved here, but as usual I am trying to prevent code sprawl by minimizing the number of unrelated changes in these PRs. If we are to minimize code change and continue the existing lazy initialization paradigm, then it ought to at least perform well in a concurrent environment by using a rigorous version of double-checked locking.

Reply:
Sure, makes sense, just something that caught my eye during the review that could perhaps be cleaned up later.